### PR TITLE
Add failing test for Given/When/Then framework

### PR DESCRIPTION
**Pull request by mattwynne**
Over to you @jowers ;)

### CODE REVIEW

**Comment by jowers** [comment]
@mattwynne I've created a fork and had a go at filling out the framework.tcl, but have hit a problem, I suspect to do with when/where files are being sourced. My checkin is https://github.com/jowers/cucumber-ruby-tcl/commit/121db8411feb8992d79229c2acd3e1f3dc0efc9f

The step definition isn't calling my new Given, When, Then Procs ( or for some reason, I'm not sourcing them), which means that when it comes to call step_definition_exists, it finds nothing, and we get a pending step.

**Comment by jowers** [comment]
Ok, I've added another commit to my fork that actually tries to source the files.

It's not finding anything though. If I source lib/cucumber/tcl/framework.tcl from a tclsh, it does find files, and calls the given, when, then procs, so there must be something different in the way it's being called from the ruby-tcl interop?

**Comment by mattwynne** [comment]
Could it just be the path it's searching? Maybe the current working dir is different?

**Comment by mattwynne** [comment]
@jowers you don't need to use a fork - you have permissions to push to this repo, so we may as well work on this branch here together. I've pulled in the changed from your fork.

8b5c3c9570e9da4a99c135aecf585dae5763093b gets us a bit closer, in that it now seems to find the `*.tcl` files and source them. I think loading any `tcl` file in the `features` directory is better anyway, since that's how Ruby Cucumber behaves - the `step_definitions` directory is just a convention.

**Comment by jowers** [comment]
Ok, I've now got the basic test passing - although travis doesn't seem too happy - is there something I've missed - previously it was finding a .travis.yml file, which now doesn't appear to be found?

**Comment by mattwynne** [comment]
Excellent! Travis is in another branch. I was waiting to merge it until I have it passing.

**Comment by mattwynne** [comment]
Build is now working!

@jowers can you just update the README to make sense now that we have this TCL step definition DSL?

Also, do you think it's worth adding any unit tests around the DSL framework code (using [tclspec](https://github.com/arthurschreiber/tclspec) or similar) to help guide future maintainers?

@barneyfisher any thoughts from you? Have you reviewed @jowers's code ;)

File: lib/cucumber/tcl/framework.tcl [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+proc execute_step_definition { step_name } {
+  # TODO: handle parameters in the regexp
+  global STEPS
+
+  foreach step $STEPS {
+    set existing_step_name [lindex $step 0]
+    set existing_step_body [lindex $step 1]
+
+    if {[regexp $existing_step_name $step_name matchresult]} {
```
Comment (mattwynne) [inline lib/cucumber/tcl/framework.tcl]: There's some duplication here - this loop and if statement is in the other method too. Can you extract out a `find_step_definition` proc or somesuch?

**Comment by jowers** [comment]
Yup, I'll have a look at simplifying that proc, and update the README. I'll also have a look at tclspec to see if it's something we could use.

I was also thinking about adding more features (e.g. a failing step), but was going to do that on a separate branch to allow us to get this stuff merged quicker, unless you think it's worth having extra tests in here now?

**Comment by mattwynne** [comment]
I think having some tests for failing steps etc is a good idea, but like you say we can add them later.

**Comment by jowers** [comment]
README updated. Will look at the code re-use and unit tests next.

@mattwynne Where would you suggest I put the unit tests?

**Comment by jowers** [comment]
I've refactored the foreach loops into a single proc and added in some unit tests - I went for tcltest in the end as it's included in the core tcl install - tclspec required some exotic packages which seemed overkill.

To allow the unit tests to be run, you have to set TEST=1 in your environment (otherwise it was trying to source step definition files, and that seemed the easiest way.

**Comment by aslakhellesoy** [comment]
I love to watch how support for a new programming language takes shape. Great work @jowers and @mattwynne !

**Comment by mattwynne** [comment]
Great work with the test @jowers! How do you run them? Can you add a new task to the Rakefile so they'll run in the build?

**Comment by jowers** [comment]
Added task to the Rakefile - tcltest - first time I've tried doing this, so let me know if I've got anything wrong :-)

**Comment by mattwynne** [comment]
This looks good to me! I'll merge it into master.

**Comment by mattwynne** [comment]
Perfect!

**Pull request merged**